### PR TITLE
Avoid chown data_dir during the upgrade process

### PR DESCRIPTION
**Pull request by kay0u**
## Problem

- *chown the world*

## Solution

- *don't chown/chmod data_dir and his content during the upgrade*
- *add a button in the config panel to launch the corresponding chown/chmod command (this my fix issue in some case I guess?)*

## PR Status

- [ ] Code finished and ready to be reviewed/tested
- [ ] The fix/enhancement were manually tested (if applicable)

## Automatic tests

Automatic tests can be triggered on https://ci-apps-dev.yunohost.org/ *after creating the PR*, by commenting "!testme", "!gogogadgetoci" or "By the power of systemd, I invoke The Great App CI to test this Pull Request!". (N.B. : for this to work you need to be a member of the Yunohost-Apps organization)

### CODE REVIEW

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

File: scripts/upgrade [hunk 1]
```diff
@@ -138,7 +138,8 @@ then
     ynh_script_progression --message="Upgrading $app..." --weight=3
 
     # Set write access for the following commands
-    chown -R $app: "$install_dir" "$data_dir"
+    find "$install_dir" -not \( -user $app -or -group www-data \) -exec chown $app:www-data {} \+
+    find "$data_dir" -not \( -user $app -or -group $app \) -exec chown $app: {} \+
```
[the remark this answers is not in the document]
Comment (Thovi98) [reply scripts/upgrade:143]: I’ve taken the habit to remove the several chown/chmod $datadir before making the upgrade, it was always successful and the upgrade went from +-25 hours to +-20 minutes ^^

[the remark this answers is not in the document]
Comment (kay0u) [reply scripts/upgrade:143]: Thanks for the feedback, I think it's a really great improvement

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

**Comment by CodeShakingSheep** [comment]
Thanks for working on that. Upgrading my nextcloud from 29.0.2 to 29.0.4 ran into a Gateway timeout (HTTP 504).
image

Although `chown` command is still running (for almost 2 hours).
image

File: scripts/install [hunk 2]
```diff
@@ -64,7 +64,8 @@ exec_occ() {
 }
 
 # Set write access for the following commands
-chown -R $app: "$install_dir" "$data_dir"
+chown -R $app:www-data "$install_dir"
+chown -R $app: "$data_dir"
```
Comment (alexAubin) [inline scripts/install:68]:
```suggestion
chown -R $app:www-data "$install_dir"
chmod og-rwx "$install_dir/config"
chown -R $app: "$data_dir"
```

(me being paranoid about compromission of www-data that would result in leaking secrets etc)

File: scripts/upgrade [hunk 3]
```diff
@@ -207,7 +207,7 @@ then
         mv "$tmpdir" "$install_dir"
 
         # Set write access for the following commands
-        chown -R $app: "$install_dir" "$data_dir"
+        chown -R $app:www-data "$install_dir"
```
Comment (alexAubin) [inline scripts/upgrade:210]:
```suggestion
        chown -R $app:www-data "$install_dir"
        chmod og-rwx "$install_dir/config"
```

(me being paranoid about compromission of www-data that would result in leaking secrets etc)

**Comment by kay0u** [comment]
@CodeShakingSheep (? with these changes or with the new stable version?) Then just wait, and don't try to restart your server or `yunohost-api`, the api timed-out but the upgrade is still running

**Comment by CodeShakingSheep** [comment]
> @CodeShakingSheep (? with these changes or with the new stable version?) Then just wait, and don't try to restart your server or `yunohost-api`, the api timed-out but the upgrade is still running

With the stable version. I didn't restart anything but my whole server just went down because of this. So, I have to restart it now.

File: scripts/install (hunk 2 above)
Comment (kay0u) [reply scripts/install:68]: Idk, do you think it's not enough?

https://github.com/YunoHost-Apps/nextcloud_ynh/blob/30b2d8123a13d315c803486d9abee3607cc2162f/scripts/install#L244

https://github.com/YunoHost-Apps/nextcloud_ynh/blob/30b2d8123a13d315c803486d9abee3607cc2162f/scripts/restore#L88

https://github.com/YunoHost-Apps/nextcloud_ynh/blob/30b2d8123a13d315c803486d9abee3607cc2162f/scripts/upgrade#L330

Comment (alexAubin) [reply scripts/install:68]: Eeeh chmod 640 will still allow www-data to read the file, so in the event that www-data gets compromised (or some funky path traversal issue is discovered or whatever) it still allows to access secrets

**Comment by kay0u** [comment]
!testme

**Comment by kay0u** [comment]
!testme

**Pull request merged**